src/channel-qvm.lisp

```diff
@@ -0,0 +1,374 @@
+    ret-qvm))
+
+
+(defun matches-rule (rule instr-qubits instr-name position)
+  "check if rule is matched by instruction data and the position of the match request"
+  (let* ((noise-predicate (predicate (noise-predicate rule)))
+         (noise-position (noise-position (noise-predicate rule))))
+    (and (funcall noise-predicate instr-qubits instr-name) (string= position noise-position))))
+
+
+(defun match-rules (rules instr-qubits instr-name position)
+  "return the operation elements for the first matching rule"
+  (loop :for rule in rules
+        :when (matches-rule rule instr-qubits instr-name position)
+          :return (operation-elements rule)))
+
+
+(defmethod apply-all-kraus-operators ((qvm channel-qvm) (instr quil:gate-application) kraus-operators)
+  "Apply all the kraus operators to the state of the system."
+  (loop :for kops :in kraus-operators
+        :do
+           (check-kraus-ops kops)
+           (apply-kraus-ops qvm instr kops)))
+
+
+(defmethod apply-kraus-ops ((qvm channel-qvm) (instr quil:gate-application) kraus-ops)
+  (let* ((amps (%trial-amplitudes qvm))
+         (r (random 1.0d0))
+         (summed-p 0.0d0)
+         (logical-qubits (quil:application-arguments instr))
+         (qubits (mapcar #'quil:qubit-index logical-qubits))) ;; this needs to be some kind of reference to operational qubits. Have another dict? 
+    (check-type amps quantum-state)
+    (loop :for kj :in kraus-ops
+          :do
+             (replace amps (amplitudes qvm))
+             (apply-matrix-operator (magicl-matrix-to-quantum-operator kj) amps (apply #'nat-tuple qubits))
+             (incf summed-p (psum #'probability amps))
+          :until (>= summed-p r))
+
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))
+    (normalize-wavefunction (amplitudes qvm))))
+
+
+(defgeneric apply-classical-readout-noisee (qvm instr)
+  (:documentation "Given a QVM and a (measurement) instruction INSTR, corrupt the readout bit according to the POVM specifications of QVM.")
+  ;; Pure state QVM has no readout noise.
+
+  (:method ((qvm pure-state-qvm) (instr quil:measurement))
+    (declare (ignore qvm instr))
+    nil)
+  ;; Readout noise only happens to the resulting classical bit (i.e.,
+  ;; it's classical noise). As such, discarding that bit doesn't
+  ;; warrant any sort of special treatment.
+  (:method ((qvm approx-qvm) (instr quil:measure))
+    (%corrupt-readout qvm instr (readout-povms qvm))
+    )
+  (:method ((qvm channel-qvm) (instr quil:measure-discard))
+    (declare (ignore qvm instr))
+    nil)
+  ;; We do have a readout bit, and we want to corrupt it.
+  (:method ((qvm channel-qvm) (instr quil:measure))
+    (%corrupt-readout qvm instr (readout-povms (noise-model qvm))))
+  ;; For compiled measurements, refer to the source of that
+  ;; instruction.
+  (:method ((qvm channel-qvm) (instr compiled-measurement))
+    (apply-classical-readout-noise qvm (source-instruction instr))))
```

Comment:
I think you can replace this `defgeneric` with a series of `(defmethod apply-classical-readout-noise ...)`, unless there is some reason you want to keep `apply-classical-readout-noise` and `apply-classical-readout-noisee` separate for debugging or whatever for now.

Reply:
nope, that was just a typo I forgot to fix :(

Reply:
I copied this function to match how readout noise is applied in `noisy-qvm.lisp`

Reply:
Makes sense. If it's just a typo and you don't want a separately-named generic function, then you probably want to split out all the `(:method ...)` forms here into their own toplevel `(defmethod ...)` forms (except for the one specialized on `pure-state-qvm` which you can drop since it is already defined in `noisy-qvm.lisp`). While you're at it, you might want to move the methods that are specialized on `approx-qvm` into the file where that class is defined. Check out the `(defmethod apply-classical-readout-noise ...)` in density-qvm.lisp for an example.

The syntax above is a convenient shorthand that lets you define a generic function and bunch of related methods all in one go, rather having a separate `defgeneric` followed by a bunch of `defmethod`s.  But if the generic function is already defined (in noisy-qvm.lisp), then this will replace its existing methods (assuming you fix the typo and change the name to match the one in noisy-qvm.lisp), which probably isn't what you want.

Reply:
Ok, sounds good. In case we end up deprecating the defgeneric in `noisy-qvm.lisp`, I'll also specify a defmethod for the `pure-state-qvm` option and move it into `qvm.lisp`, so that I don't accidentally forget to handle the `pure-state-qvm` case